src/Maps/components/GoogleMap.ts

```diff
@@ -0,0 +1,175 @@
+                rotateControl: this.props.rotateControl,
+                styles: this.getMapsStyles(),
+                minZoom: 2,
+                maxZoom: 20
+            });
+        }
+        this.setDefaultCenter(props);
+    }
+
+    private setDefaultCenter = (props: GoogleMapsProps) => {
+        const { defaultCenterLatitude, defaultCenterLongitude, fetchingData } = props;
+        if (defaultCenterLatitude && defaultCenterLongitude) {
+            this.setState({
+                center: {
+                    lat: Number(defaultCenterLatitude),
+                    lng: Number(defaultCenterLongitude)
+                }
+            });
+        } else if (!fetchingData) {
+            this.addMarkers(props.allLocations);
+        }
+    }
+
+    private addMarkers = (mapLocations?: Location[]) => {
+        this.markers = [];
+        if (mapLocations && mapLocations.length) {
+            this.bounds = new google.maps.LatLngBounds();
+            mapLocations.forEach(location => {
+                this.bounds.extend({
+                    lat: Number(location.latitude),
+                    lng: Number(location.longitude)
+                });
+                const marker = new google.maps.Marker({
+                    position: {
+                        lat: Number(location.latitude),
+                        lng: Number(location.longitude)
+                    },
+                    icon: location.url ? location.url : undefined
+                });
+                marker.addListener("click", (event: google.maps.MouseEvent) => {
+                    if (this.props.onClickMarker && location.locationAttr) {
+                        this.props.onClickMarker(event, location.locationAttr);
+                    }
+                });
+                this.markers.push(marker);
+            });
+            this.setMapOnMarkers(this.map);
+            this.setBounds(this.bounds);
+        }
+    }
+
+    private setBounds = (mapBounds?: google.maps.LatLngBounds) => {
+        setTimeout(() => {
+            if (mapBounds && this.map) {
+                try {
+                    this.map.fitBounds(mapBounds);
+                    if (!this.props.autoZoom) {
+                        this.map.setZoom(this.props.zoomLevel);
+                    }
+                } catch (error) {
+                    this.setState({ alertMessage: `Failed due to ${error.message}` });
+                }
+            }
+        }, 0);
+    }
+
+    private setMapOnMarkers = (map?: google.maps.Map) => {
+        if (this.markers && this.markers.length && map) {
+            this.markers.forEach(marker => marker.setMap(map));
+        }
+    }
+
+    private getMapsStyles(): google.maps.MapTypeStyle[] {
+        if (this.props.mapStyles && this.props.mapStyles.trim()) {
+            try {
+                return JSON.parse(this.props.mapStyles);
+            } catch (error) {
+                this.setState({ alertMessage: `invalid Maps styles, ${error}` });
```

Comment:
Your runtime error messages must be more descriptive... should at least include the source widget's friendly ID

Reply:
done